### ISSUE TITLE
Apply an IWrapSetUpTearDown attribute to all tests in a fixture

### ISSUE THREAD

**Issue by yaakov-h** [nunit/nunit#2220]
Similar to #2015, I'm trying to perform some setup and teardown code before the test's own setup and teardown code is run, and apply this to multiple test fixtures via an attribute.

I have tried a custom attribute which implements `IWrapSetUpTearDown`, but it's not called if I apply it to the entire test class, only if I apply it to each individual test. `ITestAction` is not suitable, as it does not wrap setup and teardown.

Is there any way to do this with the current NUnit framework, or if not, can we please do this in a future version of the framework? 😄 

Example:

<details>
Code:

```
using System;
using NUnit.Framework;
using NUnit.Framework.Interfaces;
using NUnit.Framework.Internal;
using NUnit.Framework.Internal.Commands;

namespace NUnitWrap
{
    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method | AttributeTargets.Interface | AttributeTargets.Assembly, Inherited = true, AllowMultiple = true)]
    sealed class SomeTestAttribute : Attribute, IWrapSetUpTearDown
    {
        public string Comment { get; set; }

        public TestCommand Wrap(TestCommand command)
            => new SomeTestCommand(command) { Comment = Comment };

        sealed class SomeTestCommand : DelegatingTestCommand
        {
            public SomeTestCommand(TestCommand innerCommand)
                : base(innerCommand)
            {
            }

            public string Comment { get; set; }

            public override TestResult Execute(TestExecutionContext context)
            {
                context.OutWriter.WriteLine($"Context: Before test - {Comment}");
                try
                {
                    return innerCommand.Execute(context);
                }
                finally
                {
                    context.OutWriter.WriteLine($"Context: After test - {Comment}");
                }
            }
        }
    }

    [SomeTest(Comment = "applied to class")]
    class WrappingTest
    {
        [Test]
        [SomeTest(Comment = "applied to method")]
        public void Test() => Assert.Pass();
    }
}
```

Output:

```
NUnit Console Runner 3.6.0
Copyright (C) 2017 Charlie Poole

Runtime Environment
   OS Version: Microsoft Windows NT 10.0.15063.0
  CLR Version: 4.0.30319.42000

Test Files
    nunitwrap.dll

=> NUnitWrap.WrappingTest.Test
Context: Before test - applied to method
Context: After test - applied to method

Run Settings
    DisposeRunners: True
    WorkDirectory: C:\Users\yaakov.smith\Documents\Visual Studio 2017\Projects\NUnitWrap\NUnitWrap\bin\Debug
    ImageRuntimeVersion: 4.0.30319
    ImageTargetFrameworkName: .NETFramework,Version=v4.6
    ImageRequiresX86: False
    ImageRequiresDefaultAppDomainAssemblyResolver: False
    NumberOfTestWorkers: 12

Test Run Summary
  Overall result: Passed
  Test Count: 1, Passed: 1, Failed: 0, Warnings: 0, Inconclusive: 0, Skipped: 0
  Start time: 2017-06-06 00:52:52Z
    End time: 2017-06-06 00:52:53Z
    Duration: 0.915 seconds

Results (nunit3) saved as TestResult.xml
```

Expected Output:

```
Context: Before test - applied to class
Context: Before test - applied to method
Context: After test - applied to method
Context: After test - applied to class
```

Alternative Expected Output:

```
Context: Before test - applied to method
Context: Before test - applied to class
Context: After test - applied to class
Context: After test - applied to method
```

Actual Output:

```
Context: Before test - applied to method
Context: After test - applied to method
```
</details>

**Comment by CharliePoole** [comment]
We have one or two similar requests, which we will eventually merge when we decide how to handle this.

Currently, the most flexible use of custom attributes is for those that are placed directly on a method. There are no interfaces like IWrapSetUpTearDownForAllTheTestsInThisFixture. :-)

Right now, only ActionAttributes may be used on a fixture and apply to all the tests. The order of execution may not be what you are looking for however.

We know this is a gap in our custom interfaces and we'll eventually add something to fill it but there is no schedule for doing it right now.

**Comment by yaakov-h** [comment]
Do you mean from a consumer's perspective, or internally?

**Comment by jnm2** [comment]
I'm interested first in the consumer perspective. Then we can match those ideas against the internals and see what makes the most sense.

**Comment by yaakov-h** [comment]
From a consumer perspective I would very much love something like:

```csharp
public interface IWrapOneTimeSetUpTearDown : ICommandWrapper
{
}
```

FWIW, I find command wrappers and delegating test commands to be a lot easier and simpler than action attributes.

It doesn't have to be the same structure - `TestCommand` etc., but I do find it very useful to be able to do the following, conceptually:

```csharp
public TestResult ActuallyRunTheTest(Test test)
{
    using (SetSomethingUp())
    {
        return inner.ActuallyRunTheTest(test);
    }
}

**Comment by jnm2** [comment]
So my need is to put an attribute on the fixture or assembly which does the equivalent of `IWrapSetUpTearDown` on each test case within the fixture, not `IWrapOneTimeSetUpTearDown` once per fixture.

The TestCommand thing is essential for obtaining the `inner` instance, so let's stay with that unless there's a different middleware pattern that's somehow compelling enough to change what we've done till now.

So the path forward looks like adding an interface per injection hook, something like `IWrapOneTimeSetUpTearDown` or `IWrapFixtureExecution`?

@nunit/framework-team do you have any thoughts? Can we enable that to be applied to an assembly and still wrap each fixture's execution? Same question for `IWrapSetUpTearDown`: can we enable that to be applied to a fixture or assembly as the equivalent of applying it to every single test method within?

**Comment by CharliePoole** [comment]
@jnm2 You are right to treat this as an important decision. We sorely need a general way to inject cross-cutting behavior so that the overly complex Action Attributes can be deprecated!

Ideally, I think this should work for __all__ attribute interfaces that apply to a test after it has been constructed. One thought I had was to define an interface that signaled to NUnit that a particular attribute should apply across all nested tests. That attribute would be in addition to any interfaces like IWrapSetUpTearDown. For example...

```C#
public class MyAttribute : ICrosscutting, IApplyToTest
{
}
```

The attribute designer would decide whether `MyAttribute` should be allowed on methods, classes or tests. If it were allowed on methods as well as classes, there is a risk that it could be used twice, hence it represents a sharp tool that can cut you!

I imagine that NUnit would save the crosscutting attributes in a stack in or parallel to the TestExecutionContext. At the time where NUnit examines a method for a set of attributes with an interface like `IApplyToTests`, it would also examine the stack of saved crosscutting attributes.

Obviously, this is rough. I suggest fleshing it out as a specification and then somebody doing a spike. Defining a new interface always runs the risk that we will stick ourselves with something unworkable, so we don't want to be hasty about it.

**Comment by yaakov-h** [comment]
@jnm2 oh yeah, sorry, I got this mixed up with #1298 in my head for some reason.

My point still stands, just with a different name.

For this one I also have the same need for `IWrapSetUpTearDown` for each test in the fixture.

From an API point of view, could we not just keep `IWrapSetUpTearDown` and have the framework search the containing class/assembly/etc. for the attribute too, not just the test method?

If you expand the details node in the OP, that's what I intuitively tried, and it just no-op'ed.

I'm not so sure an entirely new [marker?] interface is required.

**Comment by CharliePoole** [comment]
@yaakov-h In the case of `IWrapSetUpTearDown`, it's definitely possible to do this without a new attribute to indicate that the crosscutting behavior is desired. That's because the particular interface makes no sense when called for anything but an individual test case.

However, if we want to expand this to a more general approach that allows __any__ of the interfaces, which are implemented by custom attributes, to be used in this way then I think some way of distinguishing the intent is needed. So I feel we have to decide on what the eventual API will be even if this particular case can be implemented specially.

Let's think about how this would work for IApplyToTest, IApplyToContext and all the other interfaces we may want to use in a crosscutting way.

**Comment by yaakov-h** [comment]
Aren't `IApplyToTest`, `IApplyToContext` and `ITestAction` already cross-cutting? Or am I missing something?

**Comment by CharliePoole** [comment]
No, our only crosscutting feature currently is Action Attributes, implemented via ITestAction. We decided to replace it in 3.0 but never got around to a new design, hence it keeps hanging on.

IApplyToTest and IApplyToContext only function on the tests on which they appear although setting the context can have a secondary effect on other tests, since the context persists. Unlike IWrapSetUpTearDown, both of those interfaces are used on attributes that apply to classes or assemblies.

**Comment by yaakov-h** [comment]
Ah, right. I went and checked my code to discover that I had _made_ `IApplyToContext` cross-cutting by recursively iterating through the test's test children.

**Comment by jnm2** [comment]
I'm shy of adding another marker interface. Can we make `IWrapSetUpTearDown` start to apply recursively to the test tree without considering other kinds of extensibility at this time, without painting ourselves into a corner?

**Comment by CharliePoole** [comment]
@rprouse The idea of a marker attribute was intended as a straw man approach to get some discussion going. The problem I see is that we have been talking for two or three years about the need to design a new approach to cross-cutting attributes to replace ActionAttributes but nobody has yet picked it up!

It's absolutely true, as you say, that we can handle `IWrapSetUpTearDown` more expediently, without doing that design work. It's just disappointing to see it put off once again.

I'm not sure what you mean about `IWrapOneTimeSetUpTearDown`. It's not an existing interface, of course. I agree that it's meaning would be obvious, if it existed, but the implementation is not entirely obvious because of the different structure of commands that execute fixtures as compared to test cases. I once had some ideas of changing that so both levels had the same structure.

**Comment by jnm2** [comment]
@CharliePoole I'd like to work on an all-encompassing design, but for that we should probably start with an exhaustive list of all the use cases test writers might have that need framework hooks, right? Which should probably take place in an issue which is specifically about a new, all-encompassing design?

The first question I have is about `ICrosscutting` is, what attributes do we have that *shouldn't* be cross-cutting?

**Comment by CharliePoole** [comment]
Attributes are important here, but interfaces are the real thing. We know what attributes we have but we don't know what attributes users have created. However, we do know that their attributes can only work if they implement our interfaces.

So I'd start with the extensibility interfaces, think about how some of them might be used in a cross-cutting way. For some, it may not make sense at all. For others (like `IWrapSetUpTearDown`) it may be that no new marker or interface is needed.

We'll also have to deal with the fact that an attribute may implement multiple interfaces. It's possible that only one of them would be cross-cutting - unless we make it a rule that an attribute has to be either cross-cutting or non-cross-cutting across all its interfaces. Messy!

It's also possible that we don't actually have all the extensibility interfaces we need. Whew!

**Comment by gilescope** [comment]
If we implemented IApplyToContext and IDispose, it would be great if we were disposed of after the context was done. That would would be one way to give the hooks we need. For now I think I need to use the v2 ActionAttributes. In my case I'm trying to ensure a redis server is around while some of the tests are run. (I feel like testing as a whole needs rethinking from the ground up in a containerised world, but that's out of scope for this ticket.)

**Comment by CharliePoole** [comment]
@gilescope Sorry, I can think of several objects that you might want disposed, We do, of course, Dispose the TestFixture instance if it implements IDisposable. Can you be more specific?

**Comment by yaakov-h** [comment]
* Run a precondition check eagerly before a test - e.g. fail early with "you must run this test as administrator" rather than an assortment of "access is denied" errors halfway through.
* Monitor external state during the test and change the status based on that state - e.g. monitor for `TaskScheduler.UnobservedTaskException` during the test and fail the test if a faulted task went unobserved; or monitor a log file and fail the test if any errors were written.

**Comment by CharliePoole** [comment]
@jnm2 Nice list!

As a more general suggestion, I'd differentiate between cases where a test __has to__ be run in a particular way and those where it's a decision made at the point of running the test.

For stating requirements as to how a test is to run (e.g. must run in its own process) I would use an attribute specified in the test assembly. For use cases where it's a choice made by the tester at run time for an entire assembly at once, I'd expect a runner option.

**Comment by yaakov-h** [comment]
@jnm2 @CharliePoole Any progress on this?

I'm in yet another scenario where we have common setup code that we want to run before `[SetUp]` in many tests across many fixtures, but dont want to apply an attribute to each individual test method.

**Comment by Lyra2108** [comment]
I also would be interested in this. My use case is a mixture of the ParalizableAttribute and the Retry Attribute.

I would like to run our integration tests in parallel. They should be fast and running as part of the verfication process, because noone takes care about nightly test runs. Due to 25 years old code this is only possible if each test uses a separate database. I also cannot use random databases, because well 25 years old code... So there is a pool of databases, each test can grap one if it is free, run its code and check it in for the next test to be used afterwards.

The retry part is that it is faster to only reset the databases if the test failed and then try it with a clean one.

Without this change here the only way I see at the moment is to use the outdated TestAction. :/

**Comment by stevenaw** [comment]
I found the need for something like this earlier today. My use case is what was called out by the title: namely the need to apply IWrapSetUpTearDown to all tests in a fixture or assembly.

Specifically, if it helps, I'm trying to wrap each test's run within a TransactionScope without using ITestAction.

I don't have much to contribute to the design or discussion yet as I only discovered this issue just now.

**Comment by kemsky** [comment]
There are features like `AsyncLocal` which can be user to make implicit context, it is helpful when you deal with UI tests for example but you can not use `AsyncLocal` wiithout wrapper code and unfortunately `IWrapTestMethod` is the only option now.
